flink-ml-lib/src/main/java/org/apache/flink/ml/clustering/agglomerativeclustering/AgglomerativeClustering.java

```diff
@@ -0,0 +1,415 @@
+
+            for (int i = 0; i < numDataPoints; i++) {
+                output.collect(
+                        new StreamRecord<>(Row.join(inputList.get(i), Row.of(clusterIds[i]))));
+            }
+        }
+
+        private int getNextClusterId() {
+            return nextClusterId++;
+        }
+
+        private void doClustering(List<Cluster> activeClusters) {
+            int clusterOffset1 = -1, clusterOffset2 = -1;
+            boolean clusteringRunning =
+                    (k != null && activeClusters.size() > k) || (distanceThreshold != null);
+
+            while (clusteringRunning || (computeFullTree && activeClusters.size() > 1)) {
+                // Computes the distance between two clusters.
+                double minDistance = Double.MAX_VALUE;
+                for (int i = 0; i < activeClusters.size(); i++) {
+                    for (int j = i + 1; j < activeClusters.size(); j++) {
+                        double distance =
+                                computeDistanceBetweenClusters(
+                                        activeClusters.get(i), activeClusters.get(j));
+                        if (distance < minDistance) {
+                            minDistance = distance;
+                            clusterOffset1 = i;
+                            clusterOffset2 = j;
+                        }
+                    }
+                }
+
+                // Outputs the merge info.
+                Cluster cluster1 = activeClusters.get(clusterOffset1);
+                Cluster cluster2 = activeClusters.get(clusterOffset2);
+                int clusterId1 = cluster1.clusterId;
+                int clusterId2 = cluster2.clusterId;
+                output.collect(
+                        mergeInfoOutputTag,
+                        new StreamRecord<>(
+                                Tuple4.of(
+                                        Math.min(clusterId1, clusterId2),
+                                        Math.max(clusterId1, clusterId2),
+                                        minDistance,
+                                        cluster1.dataPointIds.size()
+                                                + cluster2.dataPointIds.size())));
+
+                // Merges these two clusters.
+                Cluster mergedCluster =
+                        new Cluster(
+                                getNextClusterId(), cluster1.dataPointIds, cluster2.dataPointIds);
+                activeClusters.set(clusterOffset1, mergedCluster);
+                activeClusters.remove(clusterOffset2);
+
+                // Updates cluster Ids for each data point if clustering is still running.
+                if (clusteringRunning) {
+                    int mergedClusterId = mergedCluster.clusterId;
+                    for (int dataPointId : mergedCluster.dataPointIds) {
+                        clusterIds[dataPointId] = mergedClusterId;
+                    }
+                }
+
+                clusteringRunning =
```

Comment:
How about changing the order of the logics here?
```java
                clusteringRunning =
                        (k != null && activeClusters.size() > k)
                                || (distanceThreshold != null && distanceThreshold > minDistance);

                if (clusteringRunning) {
                    int mergedClusterId = mergedCluster.clusterId;
                    for (int dataPointId : mergedCluster.dataPointIds) {
                        clusterIds[dataPointId] = mergedClusterId;
                    }
                }
```

Reply:
No, we need to merge the cluster in this round first.

Say when entering the loop, the number of active clusters is `K+1`. If we change the order as above, the number of output clusters will be `k+1`.